### DISCUSSION TITLE
Importing Hydraulics File Rules

### DISCUSSION THREAD

**Discussion by codymccoy** [HydrologicEngineeringCenter/HEC-FDA#305] [General]
I am trying to figure out what the rules are for importing hydraulics into FDA 2.0.
We want to support tif, vrt. I see in my code a file type called "flt". Do we care about that?

The user will select a folder. What do we require to be in that folder?
* 8 tif files
* 1 vrt file and 8 tif files?

**Comment by rnugent3** [comment]
I've sent three examples through DoD safe. I'd like to discuss ways that we
can allow for optimal flexibility.

**Comment by rnugent3** [comment]
During our biweekly meeting, we discussed two options - a rigorous option and a flexible option. 

### Rigorous Option 
The rigorous option involves requiring hydraulics data to be in the format of raster data exported from RAS. That means a source directory with 8 subdirectories, within which will be one .vrt with potentially many .tifs. We ask the user to point to the directory and import data by looking for a .vrt in each of the 8 subdirectories. The process is automated. 

**PROS:**
- Involves a higher level of automation, reducing the potential for user error 
- Quicker and easier, given hydraulic modeling comes from HEC-RAS (typical) 

**CONS:**
- Does not accommodate hydraulic modeling from other software, may require annoying work to get the data in 
- If there is other data saved within the hydraulics directory, we'll run into problems (e.g. two .vrt, one for depth one for WSE; or two sets of subdirectories)
- Would need to present the user the name of the subdirectory or the entire path for the user to specify the recurrence interval because the .vrt will be named the same in all subdirectories 

### Flexible Option
The flexible option involves removing the folder browser source directory and adding a file browser to each of the 8 hydraulic event rows. We could (perhaps for a later version) start with just one row and allow the user to add rows, allowing for flexibility in the required number of profiles for the compute. 

**PROS:**
- Agnostic to data organization and thus agnostic to hydraulic modeling software, no concern for messy data organization

**CONS:** 
- This option takes more work. Identify 8 files for each set of hydraulic data. One set of hydraulic data for each plan-year combination potentially means lots of pointing and clicking. 
- The more pointing and clicking, the higher chance for user error. 

![image](https://user-images.githubusercontent.com/25163165/157557228-5b4d2b54-d707-4290-9bd1-3570876ebb4b.png)

**Comment by rnugent3** [comment]
We could also go both routes. Prioritize one option for 2.0 and put out the option for 2.1. 

I think I am leaning toward prioritizing the rigorous option. It is not too much to ask the users to not mess with the data they get from RAS, and I am hoping one of the field's first impressions of 2.0 is that the software is not difficult to use.

**Reply by rnugent3** [reply]
@HenryGeorgist @Brennan1994 any thoughts about going with the rigorous option?

**Reply by HenryGeorgist** [reply]
im good with the rigorous option
though dont we want to support hdf files too?

**Reply by rnugent3** [reply]
I am glad to hear that. 

HDF files/RAS API falls under the optional approach here: https://github.com/HydrologicEngineeringCenter/HEC-FDA/discussions/301
and is not yet in the development schedule/**budget**. This discussion is specific to the required approach at the above link.

**Reply by Brennan1994** [reply]
I've added a discussion here to talk about how results come out of RAS right now. 
https://github.com/HydrologicEngineeringCenter/HEC-FDA/discussions/308

I'm fine with the rigorous approach, but it really only works well for gridded, unsteady data. Steady flow plans are slightly different in that multiple profiles are included in a single plan, which means to import steady flow, you'd have to do some unnecessary sub directory creation and organization. I don't know for sure, but I think a lot of FDA studies still use steady flow hydraulics, so that seems bad. My discussion has an example of what the data out of steady flow looks like (saved as grids)

**Comment by rnugent3** [comment]
The team made a decision about importing hydraulics, documented here: https://github.com/HydrologicEngineeringCenter/HEC-FDA/discussions/301

To import gridded hydraulics, we'll follow the rigorous approach. @codymccoy please reach out with any questions. Thanks, all!
